File: src/agent/orchestrator/recovery.ts
/**
 * Orchestrator Crash Recovery
 *
 * Handles recovery from crashes that occur during the two-phase commit pattern.
 * When the orchestrator crashes between creating a git commit and updating the task,
 * the task remains in "commit_pending" status with a pendingCommit record.
 *
 * On restart, this module:
 * 1. Finds all tasks in "commit_pending" status
 * 2. Checks if the recorded commit SHA exists in git history
 * 3. If commit exists: completes the transition to "closed"
 * 4. If commit doesn't exist: resets to "in_progress" for retry
 *
 * @module agent/orchestrator/recovery
 */

import * as FileSystem from "@effect/platform/FileSystem";
import * as Path from "@effect/platform/Path";
import { Effect } from "effect";
import { findTasksWithPendingCommit, updateTask, type TaskServiceError } from "../../tasks/service.js";
import type { Task, PendingCommit } from "../../tasks/schema.js";

export interface RecoveryResult {
  /** Tasks that were successfully closed (commit existed) */
  readonly closed: readonly Task[];
  /** Tasks that were reset to in_progress (commit didn't exist) */
  readonly reset: readonly Task[];
  /** Tasks that failed recovery */
  readonly failed: readonly { task: Task; error: string }[];
}

export interface RecoveryOptions {
  /** Path to the tasks.jsonl file */
  readonly tasksPath: string;
  /** Working directory for git operations */
  readonly cwd: string;
  /** Emit events during recovery */
  readonly emit?: (event: RecoveryEvent) => void;
}

export type RecoveryEvent =
  | { type: "recovery_start"; pendingCount: number }
  | { type: "commit_found"; taskId: string; sha: string }
  | { type: "commit_not_found"; taskId: string; sha: string | undefined }
  | { type: "task_closed"; taskId: string; sha: string }
  | { type: "task_reset"; taskId: string }
  | { type: "recovery_failed"; taskId: string; error: string }
  | { type: "recovery_complete"; result: RecoveryResult };

/**
 * Check if a commit SHA exists in the git repository.
 */
const commitExists = (
  sha: string,
  cwd: string,
): Effect.Effect<boolean, Error, never> =>
  Effect.tryPromise({
    try: async () => {
      const { execSync } = await import("node:child_process");
      try {
        // git cat-file -t returns the object type if it exists
        execSync(`git cat-file -t ${sha}`, {
          cwd,
          encoding: "utf-8",
          stdio: ["pipe", "pipe", "pipe"],
        });
        return true;
      } catch {
        return false;
      }
    },
    catch: () => new Error(`Failed to check commit ${sha}`),
  });

/**
 * Recover a single task that was in commit_pending state.
 */
const recoverTask = (
  task: Task,
  options: RecoveryOptions,
): Effect.Effect<
  { action: "closed" | "reset"; task: Task },
  TaskServiceError | Error,
  FileSystem.FileSystem | Path.Path
> =>
  Effect.gen(function* () {
    const { tasksPath, cwd, emit } = options;
    const pending = task.pendingCommit as PendingCommit;

    // Check if the commit SHA exists (if we have one)
    if (pending.sha) {
      const exists = yield* commitExists(pending.sha, cwd);

      if (exists) {
        emit?.({ type: "commit_found", taskId: task.id, sha: pending.sha });

        // Commit exists - complete the transition to closed
        const updated = yield* updateTask({
          tasksPath,
          id: task.id,
          update: {
            status: "closed",
            closeReason: "Completed by MechaCoder orchestrator (recovered)",
            pendingCommit: null,
            commits: [...task.commits, pending.sha],
          },
        });

        emit?.({ type: "task_closed", taskId: task.id, sha: pending.sha });
        return { action: "closed" as const, task: updated };
      }
    }

    // No SHA or commit doesn't exist - reset to in_progress
    emit?.({
      type: "commit_not_found",
      taskId: task.id,
      sha: pending.sha,
    });

    const updated = yield* updateTask({
      tasksPath,
      id: task.id,
      update: {
        status: "in_progress",
        pendingCommit: null,
      },
    });

    emit?.({ type: "task_reset", taskId: task.id });
    return { action: "reset" as const, task: updated };
  });

/**
 * Recover all tasks that were interrupted during two-phase commit.
 *
 * This should be called at orchestrator startup to resolve any tasks
 * that were left in "commit_pending" status due to a crash.
 */
export const recoverPendingCommits = (
  options: RecoveryOptions,
): Effect.Effect<RecoveryResult, TaskServiceError, FileSystem.FileSystem | Path.Path> =>
  Effect.gen(function* () {
    const { tasksPath, emit } = options;

    // Find all tasks with pending commits
    const pending = yield* findTasksWithPendingCommit(tasksPath);

    emit?.({ type: "recovery_start", pendingCount: pending.length });

    if (pending.length === 0) {
      const result: RecoveryResult = { closed: [], reset: [], failed: [] };
      emit?.({ type: "recovery_complete", result });
      return result;
    }

    const closed: Task[] = [];
    const reset: Task[] = [];
    const failed: { task: Task; error: string }[] = [];

    // Process each pending task
    for (const task of pending) {
      const result = yield* recoverTask(task, options).pipe(
        Effect.catchAll((error) =>
          Effect.succeed({
            action: "failed" as const,
            task,
            error: error instanceof Error ? error.message : String(error),
          }),
        ),
      );

      if (result.action === "failed") {
        emit?.({ type: "recovery_failed", taskId: task.id, error: result.error });
        failed.push({ task, error: result.error });
      } else if (result.action === "closed") {
        closed.push(result.task);
      } else {
        reset.push(result.task);
      }
    }

    const recoveryResult: RecoveryResult = { closed, reset, failed };
    emit?.({ type: "recovery_complete", result: recoveryResult });

    return recoveryResult;
  });

/**
 * Check if there are any tasks requiring recovery.
 * Useful for deciding whether to show recovery messages on startup.
 */
export const hasPendingRecovery = (
  tasksPath: string,
): Effect.Effect<boolean, TaskServiceError, FileSystem.FileSystem> =>
  findTasksWithPendingCommit(tasksPath).pipe(
    Effect.map((tasks) => tasks.length > 0),
  );

File: crates/agent/src/recovery.rs
//! Orchestrator Crash Recovery
//!
//! Handles recovery from crashes that occur during the two-phase commit pattern.
//! When the orchestrator crashes between creating a git commit and updating the task,
//! the task remains in "commit_pending" status with a pendingCommit record.
//!
//! On restart, this module:
//! 1. Finds all tasks in "commit_pending" status
//! 2. Checks if the recorded commit SHA exists in git history
//! 3. If commit exists: completes the transition to "closed"
//! 4. If commit doesn't exist: resets to "in_progress" for retry

use crate::error::AgentResult;
use crate::git::run_git;
use crate::types::OrchestratorEvent;
use serde::{Deserialize, Serialize};

/// Recovery result for a batch of tasks
#[derive(Debug, Clone, Default)]
pub struct RecoveryResult {
    /// Number of tasks successfully closed (commit existed)
    pub closed_count: u32,
    /// Number of tasks reset to in_progress (commit didn't exist)
    pub reset_count: u32,
    /// Number of tasks that failed recovery
    pub failed_count: u32,
    /// Details of each recovery action
    pub details: Vec<RecoveryAction>,
}

/// Individual recovery action
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct RecoveryAction {
    pub task_id: String,
    pub action: RecoveryActionType,
    pub commit_sha: Option<String>,
    pub error: Option<String>,
}

/// Type of recovery action taken
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum RecoveryActionType {
    Closed,
    Reset,
    Failed,
}

/// Pending commit record for a task
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct PendingCommit {
    pub sha: Option<String>,
    pub message: Option<String>,
    pub timestamp: String,
}

/// Recovery options
#[derive(Debug, Clone)]
pub struct RecoveryOptions {
    /// Working directory for git operations
    pub cwd: String,
}

impl Default for RecoveryOptions {
    fn default() -> Self {
        Self {
            cwd: ".".to_string(),
        }
    }
}

/// Check if a commit SHA exists in the git repository
pub fn commit_exists(sha: &str, cwd: &str) -> bool {
    let result = run_git(cwd, &["cat-file", "-t", sha]);
    result.exit_code == 0
}

/// Recover a single pending commit
pub fn recover_pending_commit(
    task_id: &str,
    pending: &PendingCommit,
    options: &RecoveryOptions,
    mut emit: Option<impl FnMut(OrchestratorEvent)>,
) -> RecoveryAction {
    // Check if we have a SHA to verify
    if let Some(ref sha) = pending.sha {
        if commit_exists(sha, &options.cwd) {
            // Commit exists - can safely close the task
            if let Some(ref mut emit_fn) = emit {
                emit_fn(OrchestratorEvent::RecoveryTaskClosed {
                    task_id: task_id.to_string(),
                    sha: sha.clone(),
                });
            }
            return RecoveryAction {
                task_id: task_id.to_string(),
                action: RecoveryActionType::Closed,
                commit_sha: Some(sha.clone()),
                error: None,
            };
        }
    }

    // No SHA or commit doesn't exist - reset to in_progress
    if let Some(ref mut emit_fn) = emit {
        emit_fn(OrchestratorEvent::RecoveryTaskReset {
            task_id: task_id.to_string(),
        });
    }

    RecoveryAction {
        task_id: task_id.to_string(),
        action: RecoveryActionType::Reset,
        commit_sha: pending.sha.clone(),
        error: None,
    }
}

/// Recover multiple pending commits
pub fn recover_all_pending(
    pending_tasks: &[(String, PendingCommit)],
    options: &RecoveryOptions,
    mut emit: Option<impl FnMut(OrchestratorEvent)>,
) -> RecoveryResult {
    if let Some(ref mut emit_fn) = emit {
        emit_fn(OrchestratorEvent::RecoveryStart {
            pending_count: pending_tasks.len() as u32,
        });
    }

    let mut result = RecoveryResult::default();

    for (task_id, pending) in pending_tasks {
        let action = recover_pending_commit(task_id, pending, options, emit.as_mut());

        match action.action {
            RecoveryActionType::Closed => result.closed_count += 1,
            RecoveryActionType::Reset => result.reset_count += 1,
            RecoveryActionType::Failed => result.failed_count += 1,
        }

        result.details.push(action);
    }

    if let Some(ref mut emit_fn) = emit {
        emit_fn(OrchestratorEvent::RecoveryComplete {
            closed_count: result.closed_count,
            reset_count: result.reset_count,
            failed_count: result.failed_count,
        });
    }

    result
}

/// Check if there are any pending commits that need recovery
pub fn has_pending_recovery(pending_tasks: &[(String, PendingCommit)]) -> bool {
    !pending_tasks.is_empty()
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_recovery_result_default() {
        let result = RecoveryResult::default();
        assert_eq!(result.closed_count, 0);
        assert_eq!(result.reset_count, 0);
        assert_eq!(result.failed_count, 0);
    }

    #[test]
    fn test_has_pending_recovery() {
        assert!(!has_pending_recovery(&[]));

        let pending = vec![(
            "task-1".to_string(),
            PendingCommit {
                sha: Some("abc123".to_string()),
                message: Some("Test".to_string()),
                timestamp: "2025-01-01T00:00:00Z".to_string(),
            },
        )];
        assert!(has_pending_recovery(&pending));
    }

    #[test]
    fn test_recovery_action_type() {
        let action = RecoveryAction {
            task_id: "task-1".to_string(),
            action: RecoveryActionType::Closed,
            commit_sha: Some("abc123".to_string()),
            error: None,
        };
        assert_eq!(action.action, RecoveryActionType::Closed);
    }

    #[test]
    fn test_recovery_options_default() {
        let opts = RecoveryOptions::default();
        assert_eq!(opts.cwd, ".");
    }
}
